nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/actions/SetPropertiesAction.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * (C) Copyright 2018 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Funsho David
+ */
+
+package org.nuxeo.ecm.core.bulk.actions;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ecm.core.api.CloseableCoreSession;
+import org.nuxeo.ecm.core.api.CoreInstance;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.bulk.BulkCommand;
+import org.nuxeo.ecm.core.bulk.BulkCommandHelper;
+import org.nuxeo.lib.stream.computation.ComputationContext;
+import org.nuxeo.lib.stream.computation.Record;
+import org.nuxeo.runtime.stream.BatchedProcessorTopology;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+
+/**
+ * @since 10.2
+ */
+public class SetPropertiesAction extends BatchedProcessorTopology<String> {
+
+    private static final Log log = LogFactory.getLog(SetPropertiesAction.class);
+
+    public static final String COMPUTATION_NAME = "SetProperties";
+
+    public static final String STREAM_NAME = "setProperties";
+
+    public static final String PROPERTIES_PARAMS = "properties";
+
+    protected byte[] commandAsBytes;
+
+    public SetPropertiesAction() {
+        super(COMPUTATION_NAME, STREAM_NAME);
+    }
+
+    @Override
+    protected List<String> getEntriesFromRecord(Record record) {
+        String docIds = record.getKey().split("/")[1];
+        return Arrays.asList(docIds.split("_"));
+    }
+
+    @Override
+    protected boolean flushBatch(Record record) {
+        if (commandAsBytes == null) {
+            commandAsBytes = record.getData();
+        } else if (!Arrays.equals(commandAsBytes, record.getData())) {
+            commandAsBytes = record.getData();
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected void processBatch(ComputationContext context, List<String> entries) {
+        TransactionHelper.runInTransaction(() -> {
+            for (String docId : entries) {
+                BulkCommand command = BulkCommandHelper.getBulkCommandJson(commandAsBytes);
+                Map<String, Serializable> properties = command.getParam(PROPERTIES_PARAMS);
+                try (CloseableCoreSession session = CoreInstance.openCoreSession(command.getRepository(),
```

Comment:
why are you opening a session for each document to edit ?

Reply:
Because it's a mistake :)